File: src/kernels/transport/labos/reflectance.zig
//! Purpose:
//!   Own LABOS reflectance extraction and Fourier-resolution selection.
//!
//! Physics:
//!   Converts the internal radiation field into TOA reflectance and chooses
//!   the Fourier terms required by the prepared transport input.
//!
//! Vendor:
//!   LABOS reflectance extraction stage
//!
//! Design:
//!   Reflectance extraction is separated from the orders recursion so the
//!   higher-level facade can re-use the same Fourier and source-interface
//!   policy without a monolithic solver file.
//!
//! Invariants:
//!   Scalar LABOS uses the solar column for reflectance extraction. Fourier
//!   truncation is driven by the active phase coefficients and transport input.
//!
//! Validation:
//!   See `tests/unit/transport_labos_test.zig` for reflectance and Fourier
//!   selection coverage.

const std = @import("std");
const basis = @import("basis.zig");
const common = @import("../common.zig");

fn sourceInterfaceAtLevel(
    layers: []const common.LayerInput,
    source_interfaces: []const common.SourceInterfaceInput,
    ilevel: usize,
) common.SourceInterfaceInput {
    if (source_interfaces.len == layers.len + 1 and ilevel < source_interfaces.len) {
        return source_interfaces[ilevel];
    }
    return common.sourceInterfaceFromLayers(layers, ilevel);
}

/// Purpose:
///   Resolve the highest Fourier index required by the phase coefficients.
fn maxPhaseCoefficientIndex(phase_coefficients: [basis.max_phase_coef]f64) usize {
    var max_index: usize = 0;
    for (1..basis.max_phase_coef) |idx| {
        if (@abs(phase_coefficients[idx]) > 1.0e-12) {
            max_index = idx;
        }
    }
    return max_index;
}

fn maxInterfacePhaseCoefficientIndex(
    layers: []const common.LayerInput,
    source_interfaces: []const common.SourceInterfaceInput,
    ilevel: usize,
) usize {
    const source_interface = sourceInterfaceAtLevel(layers, source_interfaces, ilevel);
    const above_max = maxPhaseCoefficientIndex(source_interface.phase_coefficients_above);
    const below_max = maxPhaseCoefficientIndex(source_interface.phase_coefficients_below);
    if (layers.len == 0 or ilevel == 0 or ilevel > layers.len - 1) return @max(above_max, below_max);
    return @max(above_max, below_max);
}

fn adjacentLayerPhaseCoefficientIndex(
    layers: []const common.LayerInput,
    ilevel: usize,
) usize {
    if (layers.len == 0) return 0;
    if (ilevel == 0) return maxPhaseCoefficientIndex(layers[0].phase_coefficients);
    if (ilevel >= layers.len) return maxPhaseCoefficientIndex(layers[layers.len - 1].phase_coefficients);
    return @max(
        maxPhaseCoefficientIndex(layers[ilevel - 1].phase_coefficients),
        maxPhaseCoefficientIndex(layers[ilevel].phase_coefficients),
    );
}

fn reuseLayerKernelIndex(
    layers: []const common.LayerInput,
    source_interface: common.SourceInterfaceInput,
    ilevel: usize,
) ?usize {
    if (layers.len == 0) return null;
    const above_index = @min(ilevel, layers.len - 1);
    if (!std.mem.eql(
        f64,
        source_interface.phase_coefficients_above[0..],
        layers[above_index].phase_coefficients[0..],
    )) {
        return null;
    }
    return above_index;
}

/// Purpose:
///   Compute TOA reflectance from the resolved LABOS internal radiation field.
pub fn calcReflectance(
    ud: []const basis.UDField,
    end_level: usize,
    geo: *const basis.Geometry,
) f64 {
    const solar_col: usize = 1;
    const view_idx = geo.viewIdx();
    return ud[end_level].U.col[solar_col].get(view_idx);
}

/// Purpose:
///   Compute the integrated reflectance using source-interface or quadrature
///   carriers.
///
/// Physics:
///   Integrates the local source term against the LABOS upwelling and
///   downwelling fields, including the zero-Fourier direct term.
///
/// Vendor:
///   `LABOS reflectance extraction`
///
/// Inputs:
///   `layers` and `source_interfaces` describe the transport grid, `ud`
///   carries the internal radiation field, and `i_fourier` selects the phase
///   term.
///
/// Outputs:
///   Returns the total reflectance contribution for the requested Fourier term.
///
/// Assumptions:
///   The carrier arrays are aligned with the transport grid and any quadrature
///   grid is already normalized to that layout.
///
/// Validation:
///   `tests/unit/transport_labos_test.zig`
pub fn calcIntegratedReflectance(
    layers: []const common.LayerInput,
    source_interfaces: []const common.SourceInterfaceInput,
    rtm_quadrature: common.RtmQuadratureGrid,
    ud: []const basis.UDField,
    end_level: usize,
    i_fourier: usize,
    geo: *const basis.Geometry,
) f64 {
    const max_phase_index = if (rtm_quadrature.isValidFor(layers.len))
        @max(maxFourierIndex(layers), maxFourierIndexQuadrature(rtm_quadrature))
    else
        @max(maxFourierIndex(layers), maxFourierIndexInterfaces(source_interfaces));
    const plm_basis = basis.FourierPlmBasis.init(i_fourier, max_phase_index, geo);
    return calcIntegratedReflectanceWithBasis(
        layers,
        source_interfaces,
        rtm_quadrature,
        ud,
        end_level,
        i_fourier,
        geo,
        &plm_basis,
        null,
        null,
    );
}

/// Purpose:
///   Compute the integrated reflectance using a caller-provided Fourier basis
///   and optional layer phase-kernel cache.
///
/// Physics:
///   Reuses the exact phase-kernel basis across interfaces and reuses already
///   materialized layer kernels when the source-interface carrier is identical
///   to the layer-above phase contract.
pub fn calcIntegratedReflectanceWithBasis(
    layers: []const common.LayerInput,
    source_interfaces: []const common.SourceInterfaceInput,
    rtm_quadrature: common.RtmQuadratureGrid,
    ud: []const basis.UDField,
    end_level: usize,
    i_fourier: usize,
    geo: *const basis.Geometry,
    plm_basis: *const basis.FourierPlmBasis,
    layer_phase_kernel_cache: ?[]const basis.PhaseKernel,
    layer_phase_kernel_valid: ?[]const bool,
) f64 {
    const solar_col: usize = 1;
    const view_idx = geo.viewIdx();
    const solar_idx = geo.n_gauss + 1;
    const view_mu = @max(geo.u[view_idx], 1.0e-12);
    var reflectance: f64 = 0.0;
    const use_rtm_quadrature = rtm_quadrature.isValidFor(layers.len);

    for (0..end_level + 1) |ilevel| {
        // DECISION:
        //   Prefer the quadrature carrier when it is aligned with the layer
        //   grid; otherwise fall back to the source-interface contract.
        const source_interface = if (use_rtm_quadrature)
            common.SourceInterfaceInput{}
        else
            sourceInterfaceAtLevel(layers, source_interfaces, ilevel);
        const source_rtm_weight = if (use_rtm_quadrature)
            rtm_quadrature.levels[ilevel].weight
        else if (source_interface.rtm_weight > 0.0 and source_interface.ksca_above > 0.0)
            source_interface.rtm_weight
        else
            source_interface.source_weight;
        const source_ksca = if (use_rtm_quadrature)
            rtm_quadrature.levels[ilevel].ksca
        else if (source_interface.rtm_weight > 0.0 and source_interface.ksca_above > 0.0)
            source_interface.ksca_above
        else
            1.0;
        if (source_rtm_weight <= 0.0 or source_ksca <= 0.0) continue;
        const phase_coefficients = if (use_rtm_quadrature)
            rtm_quadrature.levels[ilevel].phase_coefficients
        else
            source_interface.phase_coefficients_above;
        const source_max_phase_index = if (use_rtm_quadrature)
            adjacentLayerPhaseCoefficientIndex(layers, ilevel)
        else if (layers.len != 0)
            adjacentLayerPhaseCoefficientIndex(layers, ilevel)
        else
            maxInterfacePhaseCoefficientIndex(layers, source_interfaces, ilevel);
        if (i_fourier > source_max_phase_index) {
            // PARITY:
            //   DISAMAR gates an integrated-source level by the max phase
            //   order of its adjacent reduced layers, then uses the interface
            //   carrier only up to that same ceiling.
            continue;
        }

        const z = blk: {
            if (!use_rtm_quadrature) {
                if (reuseLayerKernelIndex(layers, source_interface, ilevel)) |above_index| {
                    if (layer_phase_kernel_cache) |cache| {
                        if (layer_phase_kernel_valid) |valid| {
                            const cache_index = above_index + 1;
                            if (cache_index < cache.len and cache_index < valid.len and valid[cache_index]) {
                                break :blk cache[cache_index];
                            }
                        }
                    }
                }
            }
            break :blk basis.fillZplusZminFromBasisLimited(
                i_fourier,
                phase_coefficients,
                source_max_phase_index,
                geo,
                plm_basis,
            );
        };
        var pmin_ed: f64 = 0.0;

        for (0..geo.n_gauss) |imu| {
            const mu = @max(geo.u[imu], 1.0e-12);
            const pmin = 0.25 * z.Zmin.get(view_idx, imu) / (view_mu * mu);
            pmin_ed += pmin * ud[ilevel].D.col[solar_col].get(imu);
        }

        const solar_mu = @max(geo.u[solar_idx], 1.0e-12);
        const pmin_direct = 0.25 * z.Zmin.get(view_idx, solar_idx) / (view_mu * solar_mu);
        pmin_ed += pmin_direct * ud[ilevel].E.get(solar_idx);

        var pplusst_u: f64 = 0.0;
        for (0..geo.n_gauss) |imu| {
            const mu = @max(geo.u[imu], 1.0e-12);
            const pplusst = 0.25 * z.Zplus.get(view_idx, imu) / (view_mu * mu);
            pplusst_u += pplusst * ud[ilevel].U.col[solar_col].get(imu);
        }

        // PARITY: `LabosModule::CalcReflectance` forms the level source as
        // `E * ksca * (...)`, then applies `RTMweight` in a separate reduction.
        const contribution = ud[ilevel].E.get(view_idx) *
            source_ksca *
            (pmin_ed + pplusst_u);
        reflectance += source_rtm_weight * contribution;
    }

    if (i_fourier == 0) {
        // PARITY:
        //   Keep the vendor scalar direct term in the zero-Fourier closure.
        reflectance += ud[0].E.get(view_idx) * ud[0].U.col[solar_col].get(view_idx);
    }

    return reflectance;
}

/// Purpose:
///   Return the total non-negative scattering optical depth of the layer set.
pub fn totalScatteringOpticalDepth(layers: []const common.LayerInput) f64 {
    var total: f64 = 0.0;
    for (layers) |layer| total += @max(layer.scattering_optical_depth, 0.0);
    return total;
}

fn maxFourierIndex(layers: []const common.LayerInput) usize {
    var max_index: usize = 0;
    for (layers) |layer| {
        max_index = @max(max_index, maxPhaseCoefficientIndex(layer.phase_coefficients));
    }
    return max_index;
}

fn maxFourierIndexInterfaces(source_interfaces: []const common.SourceInterfaceInput) usize {
    var max_index: usize = 0;
    for (source_interfaces) |source_interface| {
        max_index = @max(max_index, maxPhaseCoefficientIndex(source_interface.phase_coefficients_above));
        max_index = @max(max_index, maxPhaseCoefficientIndex(source_interface.phase_coefficients_below));
    }
    return max_index;
}

fn maxFourierIndexQuadrature(rtm_quadrature: common.RtmQuadratureGrid) usize {
    var max_index: usize = 0;
    for (rtm_quadrature.levels) |level| {
        if (level.weight <= 0.0 or level.ksca <= 0.0) continue;
        max_index = @max(max_index, maxPhaseCoefficientIndex(level.phase_coefficients));
    }
    return max_index;
}

/// Purpose:
///   Resolve the highest phase-coefficient index needed by the current
///   transport input.
pub fn resolvedPhaseCoefficientMax(input: common.ForwardInput) usize {
    var max_index = maxFourierIndex(input.layers);
    if (input.rtm_quadrature.isValidFor(input.layers.len)) {
        max_index = @max(max_index, maxFourierIndexQuadrature(input.rtm_quadrature));
    } else if (input.source_interfaces.len == input.layers.len + 1) {
        max_index = @max(max_index, maxFourierIndexInterfaces(input.source_interfaces));
    }
    return max_index;
}

/// Purpose:
///   Resolve the highest Fourier term needed by the active transport input.
///
/// Vendor:
///   `LABOS reflectance extraction`
///
/// Assumptions:
///   The route has already resolved its layer and source-interface contract.
///
/// Validation:
///   `tests/unit/transport_labos_test.zig`
pub fn resolvedFourierMax(input: common.ForwardInput, controls: common.RtmControls) usize {
    _ = controls;
    if (input.layers.len == 0) return 0;
    // PARITY:
    //   Near-nadir and near-normal geometries collapse to the scalar Fourier
    //   term in the vendor path.
    if ((1.0 - input.muv) < 1.0e-5 or (1.0 - input.mu0) < 1.0e-5) return 0;
    if (input.rtm_quadrature.isValidFor(input.layers.len)) {
        return maxFourierIndexQuadrature(input.rtm_quadrature);
    }
    if (input.source_interfaces.len == input.layers.len + 1) {
        return maxFourierIndexInterfaces(input.source_interfaces);
    }
    return maxFourierIndex(input.layers);
}

test "cached layer kernels preserve integrated reflectance when source interfaces mirror layers" {
    const geo = basis.Geometry.init(4, 0.58, 0.64);
    const layers = [_]common.LayerInput{
        .{
            .scattering_optical_depth = 0.22,
            .phase_coefficients = .{ 1.0, 0.56, 0.24, 0.09 } ++
                .{0.0} ** (basis.max_phase_coef - 4),
        },
        .{
            .scattering_optical_depth = 0.18,
            .phase_coefficients = .{ 1.0, 0.41, 0.17, 0.05 } ++
                .{0.0} ** (basis.max_phase_coef - 4),
        },
    };
    var ud: [3]basis.UDField = undefined;
    for (&ud) |*field| {
        field.* = .{
            .E = basis.Vec.zero(geo.nmutot),
            .U = basis.Vec2.zero(geo.nmutot),
            .D = basis.Vec2.zero(geo.nmutot),
        };
    }
    for (0..ud.len) |ilevel| {
        for (0..geo.nmutot) |imu| {
            ud[ilevel].E.set(imu, 0.8 - 0.1 * @as(f64, @floatFromInt(ilevel)));
            for (0..2) |imu0| {
                ud[ilevel].U.col[imu0].set(imu, 0.02 * @as(f64, @floatFromInt((ilevel + 1) * (imu + 1) * (imu0 + 1))));
                ud[ilevel].D.col[imu0].set(imu, 0.01 * @as(f64, @floatFromInt((ilevel + 2) * (imu + 1) * (imu0 + 1))));
            }
        }
    }

    var source_interfaces: [3]common.SourceInterfaceInput = undefined;
    common.fillSourceInterfacesFromLayers(&layers, &source_interfaces);
    const input: common.ForwardInput = .{
        .layers = &layers,
        .source_interfaces = &source_interfaces,
    };
    const i_fourier: usize = 1;
    const plm_basis = basis.FourierPlmBasis.init(i_fourier, resolvedPhaseCoefficientMax(input), &geo);
    var kernel_cache: [3]basis.PhaseKernel = undefined;
    var kernel_valid = [_]bool{false} ** 3;
    for (0..layers.len) |layer_idx| {
        kernel_cache[layer_idx + 1] = basis.fillZplusZminFromBasis(
            i_fourier,
            layers[layer_idx].phase_coefficients,
            &geo,
            &plm_basis,
        );
        kernel_valid[layer_idx + 1] = true;
    }

    const baseline = calcIntegratedReflectance(
        &layers,
        &source_interfaces,
        .{},
        &ud,
        layers.len,
        i_fourier,
        &geo,
    );
    const cached = calcIntegratedReflectanceWithBasis(
        &layers,
        &source_interfaces,
        .{},
        &ud,
        layers.len,
        i_fourier,
        &geo,
        &plm_basis,
        &kernel_cache,
        &kernel_valid,
    );

    try std.testing.expectApproxEqAbs(baseline, cached, 1.0e-12);
}

test "integrated source truncates quadrature phase kernels by adjacent layers" {
    const geo = basis.Geometry.init(4, 0.58, 0.64);
    const layer_phase = .{ 1.0, 0.18, 0.31 } ++ .{0.0} ** (basis.max_phase_coef - 3);
    const source_phase_with_tail = .{ 1.0, 0.18, 0.31, 4.0 } ++ .{0.0} ** (basis.max_phase_coef - 4);
    const source_phase_truncated = .{ 1.0, 0.18, 0.31 } ++ .{0.0} ** (basis.max_phase_coef - 3);
    const layers = [_]common.LayerInput{
        .{
            .scattering_optical_depth = 0.1,
            .phase_coefficients = layer_phase,
        },
        .{
            .scattering_optical_depth = 0.1,
            .phase_coefficients = layer_phase,
        },
    };
    var ud: [3]basis.UDField = undefined;
    for (&ud, 0..) |*field, ilevel| {
        field.* = .{
            .E = basis.Vec.zero(geo.nmutot),
            .U = basis.Vec2.zero(geo.nmutot),
            .D = basis.Vec2.zero(geo.nmutot),
        };
        for (0..geo.nmutot) |imu| {
            field.E.set(imu, 0.8 + 0.03 * @as(f64, @floatFromInt(ilevel + imu)));
            for (0..2) |col| {
                field.U.col[col].set(imu, 0.015 * @as(f64, @floatFromInt((ilevel + 1) * (imu + 2) * (col + 1))));
                field.D.col[col].set(imu, 0.011 * @as(f64, @floatFromInt((ilevel + 2) * (imu + 1) * (col + 1))));
            }
        }
    }

    const rtm_quadrature_with_tail = common.RtmQuadratureGrid{ .levels = &.{
        .{},
        .{
            .weight = 1.0,
            .ksca = 1.0,
            .phase_coefficients = source_phase_with_tail,
        },
        .{},
    } };
    const rtm_quadrature_truncated = common.RtmQuadratureGrid{ .levels = &.{
        .{},
        .{
            .weight = 1.0,
            .ksca = 1.0,
            .phase_coefficients = source_phase_truncated,
        },
        .{},
    } };
    const i_fourier: usize = 0;
    const plm_basis = basis.FourierPlmBasis.init(i_fourier, 3, &geo);

    const actual = calcIntegratedReflectanceWithBasis(
        &layers,
        &.{},
        rtm_quadrature_with_tail,
        &ud,
        layers.len,
        i_fourier,
        &geo,
        &plm_basis,
        null,
        null,
    );
    const expected = calcIntegratedReflectanceWithBasis(
        &layers,
        &.{},
        rtm_quadrature_truncated,
        &ud,
        layers.len,
        i_fourier,
        &geo,
        &plm_basis,
        null,
        null,
    );

    try std.testing.expectApproxEqAbs(expected, actual, 1.0e-12);
}

File: rust_src/forward_model/radiative_transfer/labos/reflectance.rs
use super::{
    phase_basis::{FourierPlmBasis, PhaseKernel, fill_zplus_zmin_row_from_basis_limited},
    types::{Geometry, MAX_NMUTOT, UdField},
};
use crate::forward_model::{
    optical_properties::shared::phase_functions::{self, PhaseCoefficients},
    radiative_transfer::{
        common_route,
        common_types::{
            ForwardInput, LayerInput, RadiativeTransferControls, RtmQuadratureGrid,
            SourceInterfaceInput,
        },
    },
};

struct PhaseRows {
    zplus: [f64; MAX_NMUTOT],
    zmin: [f64; MAX_NMUTOT],
}

pub fn fill_adjacent_layer_phase_max_indices(
    source_phase_max_indices: &mut [usize],
    layer_phase_max_indices: &[usize],
) {
    let nlayer = layer_phase_max_indices.len();
    debug_assert!(source_phase_max_indices.len() > nlayer);
    if nlayer == 0 {
        if let Some(first) = source_phase_max_indices.first_mut() {
            *first = 0;
        }
        return;
    }

    source_phase_max_indices[0] = layer_phase_max_indices[0];
    for ilevel in 1..nlayer {
        source_phase_max_indices[ilevel] =
            layer_phase_max_indices[ilevel - 1].max(layer_phase_max_indices[ilevel]);
    }
    source_phase_max_indices[nlayer] = layer_phase_max_indices[nlayer - 1];
}

pub fn calc_reflectance(ud: &[UdField], end_level: usize, geometry: &Geometry) -> f64 {
    let solar_col = 1;
    let view_idx = geometry.view_idx();
    ud[end_level].u.col[solar_col].get(view_idx)
}

pub fn calc_reflectance_tangent(
    ud_tangent: &[UdField],
    end_level: usize,
    geometry: &Geometry,
) -> f64 {
    calc_reflectance(ud_tangent, end_level, geometry)
}

pub fn calc_integrated_reflectance(
    layers: &[LayerInput],
    source_interfaces: &[SourceInterfaceInput],
    rtm_quadrature: &RtmQuadratureGrid,
    ud: &[UdField],
    end_level: usize,
    i_fourier: usize,
    geometry: &Geometry,
) -> f64 {
    let max_phase_index = if rtm_quadrature.is_valid_for(layers.len()) {
        max_fourier_index(layers).max(max_fourier_index_quadrature(rtm_quadrature))
    } else {
        max_fourier_index(layers).max(max_fourier_index_interfaces(source_interfaces))
    };
    let plm_basis = FourierPlmBasis::init(i_fourier, max_phase_index, geometry);
    calc_integrated_reflectance_with_basis(
        layers,
        source_interfaces,
        rtm_quadrature,
        ud,
        end_level,
        i_fourier,
        geometry,
        &plm_basis,
        None,
        None,
        None,
    )
}

#[allow(clippy::too_many_arguments)]
pub fn calc_integrated_reflectance_with_basis(
    layers: &[LayerInput],
    source_interfaces: &[SourceInterfaceInput],
    rtm_quadrature: &RtmQuadratureGrid,
    ud: &[UdField],
    end_level: usize,
    i_fourier: usize,
    geometry: &Geometry,
    plm_basis: &FourierPlmBasis,
    adjacent_layer_phase_max_indices: Option<&[usize]>,
    layer_phase_kernel_cache: Option<&[PhaseKernel]>,
    layer_phase_kernel_valid: Option<&[bool]>,
) -> f64 {
    let solar_col = 1;
    let view_idx = geometry.view_idx();
    let solar_idx = geometry.n_gauss + 1;
    let view_mu = geometry.u[view_idx].max(1.0e-12);
    let use_rtm_quadrature = rtm_quadrature.is_valid_for(layers.len());
    let mut reflectance = 0.0;

    for (ilevel, level) in ud[..=end_level].iter().copied().enumerate() {
        let source = level_source(
            layers,
            source_interfaces,
            rtm_quadrature,
            ilevel,
            use_rtm_quadrature,
        );
        let Some((source_rtm_weight, source_ksca, phase_coefficients)) = source else {
            continue;
        };

        let source_max_phase_index = source_phase_max_index(
            layers,
            source_interfaces,
            adjacent_layer_phase_max_indices,
            ilevel,
            use_rtm_quadrature,
        );
        if i_fourier > source_max_phase_index {
            continue;
        }

        let phase_rows = phase_rows_for_level(
            layers,
            source_interfaces,
            i_fourier,
            geometry,
            plm_basis,
            layer_phase_kernel_cache,
            layer_phase_kernel_valid,
            ilevel,
            source_max_phase_index,
            phase_coefficients,
            use_rtm_quadrature,
        );

        let mut pmin_ed = 0.0;
        for imu in 0..geometry.n_gauss {
            let mu = geometry.u[imu].max(1.0e-12);
            let pmin = (0.25 * phase_rows.zmin[imu] / view_mu) / mu;
            pmin_ed += pmin * level.d.col[solar_col].data[imu];
        }

        let solar_mu = geometry.u[solar_idx].max(1.0e-12);
        let pmin_direct = (0.25 * phase_rows.zmin[solar_idx] / view_mu) / solar_mu;
        pmin_ed += pmin_direct * level.e.data[solar_idx];

        let mut pplusst_u = 0.0;
        for imu in 0..geometry.n_gauss {
            let mu = geometry.u[imu].max(1.0e-12);
            let pplusst = (0.25 * phase_rows.zplus[imu] / view_mu) / mu;
            pplusst_u += pplusst * level.u.col[solar_col].data[imu];
        }

        reflectance +=
            source_rtm_weight * level.e.data[view_idx] * source_ksca * (pmin_ed + pplusst_u);
    }

    if i_fourier == 0 {
        // Vendor LABOS keeps the scalar direct term in the zero-Fourier closure.
        reflectance += ud[0].e.get(view_idx) * ud[0].u.col[solar_col].get(view_idx);
    }
    reflectance
}

fn level_source(
    layers: &[LayerInput],
    source_interfaces: &[SourceInterfaceInput],
    rtm_quadrature: &RtmQuadratureGrid,
    ilevel: usize,
    use_rtm_quadrature: bool,
) -> Option<(f64, f64, PhaseCoefficients)> {
    if use_rtm_quadrature {
        let level = &rtm_quadrature.levels[ilevel];
        return (level.weight > 0.0 && level.ksca > 0.0).then_some((
            level.weight,
            level.ksca,
            level.phase_coefficients,
        ));
    }

    let source_interface = source_interface_at_level(layers, source_interfaces, ilevel);
    let (source_rtm_weight, source_ksca) =
        if source_interface.rtm_weight > 0.0 && source_interface.ksca_above > 0.0 {
            (source_interface.rtm_weight, source_interface.ksca_above)
        } else {
            (source_interface.source_weight, 1.0)
        };
    (source_rtm_weight > 0.0 && source_ksca > 0.0).then_some((
        source_rtm_weight,
        source_ksca,
        source_interface.phase_coefficients_above,
    ))
}

#[allow(clippy::too_many_arguments)]
fn phase_rows_for_level(
    layers: &[LayerInput],
    source_interfaces: &[SourceInterfaceInput],
    i_fourier: usize,
    geometry: &Geometry,
    plm_basis: &FourierPlmBasis,
    layer_phase_kernel_cache: Option<&[PhaseKernel]>,
    layer_phase_kernel_valid: Option<&[bool]>,
    ilevel: usize,
    source_max_phase_index: usize,
    phase_coefficients: PhaseCoefficients,
    use_rtm_quadrature: bool,
) -> PhaseRows {
    let view_idx = geometry.view_idx();
    if !use_rtm_quadrature {
        let source_interface = source_interface_at_level(layers, source_interfaces, ilevel);
        if let Some(above_index) = reuse_layer_kernel_index(layers, source_interface, ilevel)
            && let (Some(cache), Some(valid)) = (layer_phase_kernel_cache, layer_phase_kernel_valid)
        {
            let cache_index = above_index + 1;
            if cache_index < cache.len() && cache_index < valid.len() && valid[cache_index] {
                return row_from_cached_kernel(&cache[cache_index], view_idx);
            }
        }
    }

    let row = fill_zplus_zmin_row_from_basis_limited(
        i_fourier,
        phase_coefficients,
        source_max_phase_index,
        geometry,
        plm_basis,
        view_idx,
    );
    PhaseRows {
        zplus: row.zplus,
        zmin: row.zmin,
    }
}

fn row_from_cached_kernel(kernel: &PhaseKernel, row_index: usize) -> PhaseRows {
    let n = kernel.zplus.n;
    let mut rows = PhaseRows {
        zplus: [0.0; MAX_NMUTOT],
        zmin: [0.0; MAX_NMUTOT],
    };
    let row_offset = row_index * n;
    rows.zplus[..n].copy_from_slice(&kernel.zplus.data[row_offset..row_offset + n]);
    rows.zmin[..n].copy_from_slice(&kernel.zmin.data[row_offset..row_offset + n]);
    rows
}

fn source_phase_max_index(
    layers: &[LayerInput],
    source_interfaces: &[SourceInterfaceInput],
    adjacent_layer_phase_max_indices: Option<&[usize]>,
    ilevel: usize,
    use_rtm_quadrature: bool,
) -> usize {
    if let Some(indices) = adjacent_layer_phase_max_indices {
        return indices[ilevel];
    }
    if use_rtm_quadrature || !layers.is_empty() {
        return adjacent_layer_phase_coefficient_index(layers, ilevel);
    }
    max_interface_phase_coefficient_index(layers, source_interfaces, ilevel)
}

fn source_interface_at_level(
    layers: &[LayerInput],
    source_interfaces: &[SourceInterfaceInput],
    ilevel: usize,
) -> SourceInterfaceInput {
    if source_interfaces.len() == layers.len() + 1 && ilevel < source_interfaces.len() {
        return source_interfaces[ilevel].clone();
    }
    common_route::source_interface_from_layers(layers, ilevel)
}

fn max_phase_coefficient_index(phase_coefficients: &PhaseCoefficients) -> usize {
    phase_functions::max_phase_coefficient_index(phase_coefficients)
}

fn max_interface_phase_coefficient_index(
    layers: &[LayerInput],
    source_interfaces: &[SourceInterfaceInput],
    ilevel: usize,
) -> usize {
    let source_interface = source_interface_at_level(layers, source_interfaces, ilevel);
    max_phase_coefficient_index(&source_interface.phase_coefficients_above).max(
        max_phase_coefficient_index(&source_interface.phase_coefficients_below),
    )
}

fn adjacent_layer_phase_coefficient_index(layers: &[LayerInput], ilevel: usize) -> usize {
    if layers.is_empty() {
        return 0;
    }
    if ilevel == 0 {
        return max_phase_coefficient_index(&layers[0].phase_coefficients);
    }
    if ilevel >= layers.len() {
        return max_phase_coefficient_index(&layers[layers.len() - 1].phase_coefficients);
    }
    max_phase_coefficient_index(&layers[ilevel - 1].phase_coefficients).max(
        max_phase_coefficient_index(&layers[ilevel].phase_coefficients),
    )
}

fn reuse_layer_kernel_index(
    layers: &[LayerInput],
    source_interface: SourceInterfaceInput,
    ilevel: usize,
) -> Option<usize> {
    if layers.is_empty() {
        return None;
    }
    let above_index = ilevel.min(layers.len() - 1);
    (source_interface.phase_coefficients_above == layers[above_index].phase_coefficients)
        .then_some(above_index)
}

pub fn total_scattering_optical_depth(layers: &[LayerInput]) -> f64 {
    layers
        .iter()
        .map(|layer| layer.scattering_optical_depth.max(0.0))
        .sum()
}

fn max_fourier_index(layers: &[LayerInput]) -> usize {
    layers
        .iter()
        .map(|layer| max_phase_coefficient_index(&layer.phase_coefficients))
        .max()
        .unwrap_or(0)
}

fn max_fourier_index_interfaces(source_interfaces: &[SourceInterfaceInput]) -> usize {
    source_interfaces
        .iter()
        .map(|source_interface| {
            max_phase_coefficient_index(&source_interface.phase_coefficients_above).max(
                max_phase_coefficient_index(&source_interface.phase_coefficients_below),
            )
        })
        .max()
        .unwrap_or(0)
}

fn max_fourier_index_quadrature(rtm_quadrature: &RtmQuadratureGrid) -> usize {
    rtm_quadrature
        .levels
        .iter()
        .filter(|level| level.weight > 0.0 && level.ksca > 0.0)
        .map(|level| max_phase_coefficient_index(&level.phase_coefficients))
        .max()
        .unwrap_or(0)
}

pub fn resolved_phase_coefficient_max(input: &ForwardInput) -> usize {
    let mut max_index = max_fourier_index(&input.layers);
    if input.rtm_quadrature.is_valid_for(input.layers.len()) {
        max_index = max_index.max(max_fourier_index_quadrature(&input.rtm_quadrature));
    } else if input.source_interfaces.len() == input.layers.len() + 1 {
        max_index = max_index.max(max_fourier_index_interfaces(&input.source_interfaces));
    }
    max_index
}

pub fn resolved_fourier_max(input: &ForwardInput, controls: RadiativeTransferControls) -> usize {
    if input.layers.is_empty() {
        return 0;
    }
    if (1.0 - input.muv) < 1.0e-5 || (1.0 - input.mu0) < 1.0e-5 {
        return 0;
    }
    let resolved_max = if input.rtm_quadrature.is_valid_for(input.layers.len()) {
        max_fourier_index_quadrature(&input.rtm_quadrature)
    } else if input.source_interfaces.len() == input.layers.len() + 1 {
        max_fourier_index_interfaces(&input.source_interfaces)
    } else {
        max_fourier_index(&input.layers)
    };
    controls
        .performance_thresholds
        .capped_fourier_max(resolved_max)
}
